src/DotNetty.Common/Utilities/TaskEx.cs

```diff
@@ -72,14 +77,84 @@ public static void LinkOutcome(this Task task, TaskCompletionSource taskCompleti
                     taskCompletionSource.TryUnwrap(task.Exception);
                     break;
                 default:
-                    task.ContinueWith(
-                        LinkOutcomeContinuationAction,
-                        taskCompletionSource,
-                        TaskContinuationOptions.ExecuteSynchronously);
+                    task.ContinueWith(LinkOutcomeTcs, taskCompletionSource, TaskContinuationOptions.ExecuteSynchronously);
+                    break;
+            }
+        }
+        
+        static readonly Action<Task, object> LinkOutcomePromise = (t, promise) =>
+        {
+            switch (t.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    ((IPromise)promise).TryComplete();
+                    break;
+                case TaskStatus.Canceled:
+                    ((IPromise)promise).TrySetCanceled();
+                    break;
+                case TaskStatus.Faulted:
+                    ((IPromise)promise).TryUnwrap(t.Exception);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        };
+        
+        public static void LinkOutcome(this Task task, IPromise promise)
+        {
+            switch (task.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    promise.TryComplete();
+                    break;
+                case TaskStatus.Canceled:
+                    promise.TrySetCanceled();
+                    break;
+                case TaskStatus.Faulted:
+                    promise.TryUnwrap(task.Exception);
+                    break;
+                default:
+                    task.ContinueWith(LinkOutcomePromise, promise, TaskContinuationOptions.ExecuteSynchronously);
                     break;
             }
         }
 
+        public static async void LinkOutcome(this ValueTask future, IPromise promise)
+        {
+            try
+            {
+                await future;
```

Comment:
AbstractPromise now can flow both execution and synchronization context. What is our context management story? 
* Can we use await in dotnetty
* If we can, should it always be ConfigureAwait false?
* Should we suppress execution context?

Reply:
if you do `.ConfigureAwait(false)`, you would erase notion of custom task scheduler that dispatches continuation to this same event loop.

Reply:
So in case of ValueTask it's IValueTaskSource implementation responsibility to execute continuation. That implementation presumably not aware of any task scheduler.